node-repository/src/main/java/com/yahoo/vespa/hosted/provision/autoscale/QuestMetricsDb.java

```diff
@@ -55,10 +57,11 @@
 
     private final Clock clock;
     private final String dataDir;
-    private final CairoEngine engine;
+    private final AtomicReference<CairoEngine> engine = new AtomicReference<>();
+    // TODO Is this safe, or even possible. ThreadLocal members should be static
```

Comment:
There's only ever one instance of this class so it doesn't matter, but having a threadlocal per instance just means you'll have one threadlocal item per thread+instance. Here the threadlocal is just to create fewer instances, there's no constraint that there must be max one per thread.

Reply:
Ok, I concluded that was the case too. However I do not think that you want to use thread locals for this. The SqlCompilers needs to be closed. I assume prior to the engine too, but that you are not able to with thread locals.

Reply:
I was advised to do it by one of the QuestDb authors, but you are right they have a close() though. But have we seen any problems related to closing?

Reply:
I have not looked in detail at the core dumps so I can not say. I guess we need to dive in.